File: src/range.rs
use crate::{direction::Direction, position::Position};
use std::cmp::{max, min};

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct Range {
    anchor: Position,
    cursor: Position,
    target_column: Option<usize>,
}

impl Range {
    pub fn anchor(&self) -> &Position {
        &self.anchor
    }

    pub fn cursor(&self) -> &Position {
        &self.cursor
    }

    pub fn target_column(&self) -> Option<usize> {
        self.target_column
    }

    pub fn start(&self) -> &Position {
        min(&self.anchor, &self.cursor)
    }

    pub fn end(&self) -> &Position {
        max(&self.anchor, &self.cursor)
    }

    pub fn direction(&self) -> Direction {
        if self.is_forward() {
            Direction::Forward
        } else {
            Direction::Backward
        }
    }

    pub fn is_forward(&self) -> bool {
        self.anchor <= self.cursor
    }

    pub fn is_backward(&self) -> bool {
        self.anchor > self.cursor
    }

    pub fn is_reduced(&self) -> bool {
        self.anchor == self.cursor
    }

    pub fn is_overlapping(&self, other: &Self) -> bool {
        self.start() <= other.end() && other.start() <= self.end()
    }

    pub fn with_target_column(&self, target_column: usize) -> Option<Self> {
        if self.cursor.column < target_column {
            Some(Self {
                target_column: Some(target_column),
                ..self.clone()
            })
        } else {
            None
        }
    }

    pub fn without_target_column(&self) -> Self {
        Self {
            target_column: None,
            ..self.clone()
        }
    }

    pub fn flip(&self) -> Self {
        if self.is_reduced() {
            self.clone()
        } else {
            Self {
                anchor: self.cursor.clone(),
                cursor: self.anchor.clone(),
                target_column: None,
            }
        }
    }

    pub fn flip_forward(&self) -> Self {
        if self.is_backward() {
            self.flip()
        } else {
            self.clone()
        }
    }

    pub fn flip_backward(&self) -> Self {
        if self.is_forward() {
            self.flip()
        } else {
            self.clone()
        }
    }

    pub fn reduce(&self) -> Self {
        Self {
            anchor: self.cursor.clone(),
            ..self.clone()
        }
    }

    pub fn merge(&self, other: &Self) -> Self {
        match (self.is_forward(), other.is_forward()) {
            (true, true) => {
                // Forward
                let anchor = min(self.anchor.clone(), other.anchor.clone());
                let (cursor, target_column) = if self.cursor > other.cursor {
                    (self.cursor.clone(), self.target_column)
                } else {
                    (other.cursor.clone(), other.target_column)
                };
                Self {
                    anchor,
                    cursor,
                    target_column,
                }
            }
            (false, false) => {
                // Backward
                let anchor = max(self.anchor.clone(), other.anchor.clone());
                let (cursor, target_column) = if self.cursor < other.cursor {
                    (self.cursor.clone(), self.target_column)
                } else {
                    (other.cursor.clone(), other.target_column)
                };
                Self {
                    anchor,
                    cursor,
                    target_column,
                }
            }
            _ => {
                // Mixed
                self.merge(&other.flip())
            }
        }
    }

    pub fn is_valid(&self) -> bool {
        match self.target_column {
            None => true,
            Some(target_column) => self.cursor.column < target_column,
        }
    }
}

impl From<Position> for Range {
    fn from(position: Position) -> Self {
        Self {
            anchor: position.clone(),
            cursor: position.clone(),
            target_column: None,
        }
    }
}

impl From<(Position, Position)> for Range {
    fn from((anchor, cursor): (Position, Position)) -> Self {
        Self {
            anchor,
            cursor,
            target_column: None,
        }
    }
}

impl From<Range> for (Position, Position) {
    fn from(range: Range) -> Self {
        (range.anchor, range.cursor)
    }
}

File: src/Indigo/Core/Range.hs
{-# LANGUAGE NoFieldSelectors #-}
{-# LANGUAGE ViewPatterns #-}

{-# OPTIONS_GHC -Wno-name-shadowing #-}

module Indigo.Core.Range
  ( Range

    -- * Create
  , fromPosition
  , fromPositions

    -- * Query
  , anchor
  , cursor
  , targetColumn
  , isForward
  , isBackward
  , isReduced
  , isOverlapping

    -- * Modify
  , mapTargetColumn
  , forgetTargetColumn
  , flip
  , flipForward
  , flipBackward
  , reduce
  , merge

    -- * Consume
  , toPositions

    -- * Internal
  , isValid
  , isTargetColumnGreaterThanCursorColumn
  )
where

import Data.Default.Class (Default (..))
import Indigo.Core.Position (Position (..))
import Prelude hiding (flip)

data Range = Range
  { anchor :: !Position
  , cursor :: !Position
  , targetColumn :: !(Maybe Word)
  }
  deriving stock (Show, Eq)

instance Default Range where
  def :: Range
  def = fromPosition def

fromPosition :: Position -> Range
fromPosition p = fromPositions p p

fromPositions :: Position -> Position -> Range
fromPositions anchor cursor =
  Range
    { anchor
    , cursor
    , targetColumn = Nothing
    }

anchor :: Range -> Position
anchor r = r.anchor

cursor :: Range -> Position
cursor r = r.cursor

targetColumn :: Range -> Maybe Word
targetColumn r = r.targetColumn

isForward :: Range -> Bool
isForward r = r.anchor <= r.cursor

isBackward :: Range -> Bool
isBackward r = r.anchor > r.cursor

isReduced :: Range -> Bool
isReduced r = r.anchor == r.cursor

isOverlapping :: Range -> Range -> Bool
isOverlapping (flipForward -> r1) (flipForward -> r2) =
     (r1.anchor <= r2.anchor && r2.anchor <= r1.cursor)
  || (r2.anchor <= r1.anchor && r1.anchor <= r2.cursor)

mapTargetColumn :: (Maybe Word -> Maybe Word) -> Range -> Maybe Range
mapTargetColumn f range =
  guarded isValid range{ targetColumn = f range.targetColumn }

forgetTargetColumn :: Range -> Range
forgetTargetColumn r = r{ targetColumn = Nothing }

flip :: Range -> Range
flip r =
  Range
    { anchor = r.cursor
    , cursor = r.anchor
    , targetColumn = Nothing
    }

flipForward :: Range -> Range
flipForward r =
  if isBackward r
    then flip r
    else r

flipBackward :: Range -> Range
flipBackward r =
  if isForward r
    then flip r
    else r

reduce :: Range -> Range
reduce r = r{ anchor = r.cursor }

merge :: Range -> Range -> Range
merge r1 r2 =
  case (isForward r1, isForward r2) of
    (True, True) ->
      -- Forward
      Range{ anchor, cursor, targetColumn }
      where
      anchor = min r1.anchor r2.anchor
      (cursor, targetColumn) =
        if r1.cursor > r2.cursor
          then (r1.cursor, r1.targetColumn)
          else (r2.cursor, r2.targetColumn)
    (False, False) ->
      -- Backward
      Range{ anchor, cursor, targetColumn }
      where
      anchor = max r1.anchor r2.anchor
      (cursor, targetColumn) =
        if r1.cursor < r2.cursor
          then (r1.cursor, r1.targetColumn)
          else (r2.cursor, r2.targetColumn)
    _ ->
      -- Mixed
      merge r1 (flip r2)

toPositions :: Range -> (Position, Position)
toPositions r = (r.anchor, r.cursor)

toRawParts :: Range -> (Position, Position, Maybe Word)
toRawParts r = (r.anchor, r.cursor, r.targetColumn)

isValid :: Range -> Bool
isValid r = isTargetColumnGreaterThanCursorColumn r

isTargetColumnGreaterThanCursorColumn :: Range -> Bool
isTargetColumnGreaterThanCursorColumn r =
  maybe True (r.cursor.column <) r.targetColumn
